### ISSUE TITLE
flag parsing bug or user error?

### ISSUE THREAD

**Issue by krancour** [spf13/cobra#1398]
If I have two (unrelated commands) that make use of flags that coincidentally have the same name, and I bind them both to the same variable (because only one command+flag or the other can be used, so why not?) I see some very unexpected behavior.

In this minimal example, I have both `hello` and `goodbye` commands. Both accept a `--name` flag that is bound to a `flagName` variable. The only real difference is flags for these two commands have different default values:

```
package main

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
)

var flagName string

var rootCmd = &cobra.Command{
	Use:           "app",
	SilenceUsage:  true,
	SilenceErrors: true,
}

var helloCmd = &cobra.Command{
	Use: "hello",
	Run: func(*cobra.Command, []string) {
		fmt.Printf("Hello, %s.\n", flagName)
	},
}

var goodbyeCmd = &cobra.Command{
	Use: "goodbye",
	Run: func(*cobra.Command, []string) {
		fmt.Printf("Goodbye, %s.\n", flagName)
	},
}

func init() {
	rootCmd.AddCommand(
		helloCmd,
		goodbyeCmd,
	)

	helloCmd.Flags().StringVarP(
		&flagName,
		"name",
		"n",
		"krancour",
		"Who to say hello to",
	)

	goodbyeCmd.LocalFlags().StringVarP(
		&flagName,
		"name",
		"n",
		"weirdo",
		"Who to say goodbye to",
	)
}

func main() {
	if err := rootCmd.Execute(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
}
```

When you ask for help on the `hello` command, you see exactly what you'd expect:

```
$ go run main.go hello -h
Usage:
  app hello [flags]

Flags:
  -h, --help          help for hello
  -n, --name string   Who to say hello to (default "krancour")
```

But when you run it.....

```
$ go run foo.go hello
Hello, weirdo.
```

Clearly, the default value for the _other_ command has been applied.

My assumption is that this is a bug _or_ that my usage is way, way, way off the mark. If the latter is the case, then it would seem there are nuances at work here that aren't covered in the docs.

Any help is greatly appreciated!

**Comment by marckhouzam** [comment]
Hi @krancour.  Basically you shouldn't reuse the same variable for different flags. Please see #1375 or #1047.

**Comment by krancour** [comment]
@marckhouzam thanks for the speedy response.

fwiw, the docs really do create an impression that only flags applicable to the target command are parsed. That might be worth looking at.

In practice, how do people deal with this limitation? If I had 20 different unrelated commands that all accept a flag with the same name (but perhaps doesn't have the same _meaning_ in all contexts), am I really supposed to create package-scoped variables for that? That doesn't seem as if it scales well. Is there a pattern or technique that people use to deal with this?

**Comment by krancour** [comment]
I'm going to close this because I didn't do an exhaustive search of open issues before I opened this (sorry) and it seems it's, more or less, a dupe.

I did consult some team mates who have had prior experience with cobra and the pattern that they shared with me is a struct per command, with flags bound to member of those structs. If find that to be a nice way of organizing things, but if your CLI has a lot of commands, it's also a lot of unnecessary allocation of a lot of structs that won't ever be used. 🤷‍♂️  Not the end of the world, but not my style. In the end, I'm not using cobra.
